viewing_party/party.py

```diff
@@ -1,23 +1,165 @@
+        for data in movie_data:
+            rating_sum += data["rating"]
+            rating_count += 1
+
+    if len(user_data["watched"]) >= 1:
+        average_rating = rating_sum / rating_count
+        return average_rating
+    else:
+        return 0
+
+def get_most_watched_genre(user_data):
+    genre_counter = {}
+
+    for list, movie_data in user_data.items():
+        if list == "watched":
+            for data in movie_data:
+                genre = data["genre"]
+                if genre in genre_counter:
+                    genre_counter[genre] += 1
+                else:
+                    genre_counter[genre] = 1
+
+    if len(genre_counter) == 0:
+        return None
+    else:
+        favorite_genre = max(genre_counter, key=genre_counter.get)
+        return favorite_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    movies_watched = []
+    movies_friends_watched = []
+    unique_watched = []
+
+    for list, movie_data in user_data.items():
+        for movie_detail in user_data["watched"]:
+            movies_watched.append(movie_detail)
+        for friends_watched_movies in user_data["friends"]:
+            for friends_movie_details in friends_watched_movies["watched"]:
+                movies_friends_watched.append(friends_movie_details)
+
+    for movie in movies_watched:
+        if movie in movies_friends_watched:
+            continue
+        elif movie not in movies_friends_watched and movie not in unique_watched:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    movies_watched = []
+    movies_friends_watched = []
+    friends_unique_watched = []
+
+    for list, movie_data in user_data.items():
+        for movie_detail in user_data["watched"]:
+            movies_watched.append(movie_detail)
+        for friends_watched_movies in user_data["friends"]:
+            for friends_movie_details in friends_watched_movies["watched"]:
+                movies_friends_watched.append(friends_movie_details)
+
+    for movie in movies_friends_watched:
+        if movie in movies_watched:
+            continue
+        elif movie not in movies_watched and movie not in friends_unique_watched:
+            friends_unique_watched.append(movie)
+
+    return friends_unique_watched
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    recommendation = []
+    user_has_not_watched = get_friends_unique_watched(user_data)
+    subscribed = []
+
+    for services in user_data["subscriptions"]:
+        subscribed.append(services)
+
+    for movie in user_has_not_watched:
+        if movie["host"] in subscribed:
+            recommendation.append(movie)

```

Comment:
instead of doing two separate for loops, you can combine them like 

```python
for movie in user_has_not_watched:
        if movie['host'] in user_data['subscriptions']:
            recommendation.append(movie)

    return reommendation
```